website/source/docs/agent/options.html.md

```diff
@@ -1072,6 +1072,13 @@ default will automatically work with some tooling.
         Configures the Retry duration expressed in seconds, default value is
         600, ie: 10 minutes.
 
+    * <a name="dns_use_cache"></a><a href="dns_use_cache">`use_cache`</a> - When set to true, DNS resolution will use the agent cache described
+      in [agent caching](/api/index.html#agent-caching). This setting affects all service and prepared queries DNS requests.
+
+    * <a name="dns_cache_max_age"></a><a href="dns_cache_max_age">`dns_max_age`</a> - When [use_cache](#dns_use_cache) is enabled, the agent
```

Comment:
Please explain the difference between max_stale_duration and this setting. Does it make sense to merge the 2?

Reply:
`max_stale_duration` is the max server follower lag from the leader. `dns_cache_max_age` and its HTTP equivalent `&max-age=` is the max time a cached entry can be used before it is fetched again from a server.
I agree that these two settings could be combined so the clients would only worry about the results overall staleness without having to mess with the tidbits of different levels of cache in Consul.
However the HTTP API already makes this distinction with `&cached`, `&max-age=` vs `discovery_max_stale`. It would be misleading to have the DNS interface behave differently IMO.
Maybe this is something that could be thought about in another PR ?

Reply:
I think its just a point where we need to be very clear in the docs what the config entries do.

`dns_config.allow_stale` - allows non-leader servers to handle the RPC requests.
`dns_config.max_stale` - controls how stale the data on the non-leader is allowed to be.

`dns_config.use_cache` - allow caching of RPC responses at the agent level (even on servers this would be useful as cross-dc RPCs could be cached).
`dns_config.cache_max_age` - controls the lifetime of the RPC results agent cache entries.

Both of these are useful to improve DNS latency/performance but depending on your environment you may want both on, only one or none at all.

Reply:
See my other comments about `allow_stale` being implied by using the cache. The max stale duration is still respected though.